Failing test: test_running_test_pretrained_model_ddp

I think this is another problem stemming from the fact that we don't have a way to pass data back from `torch.multiprocessing.spawn`. Needs more investigation.

```
def test_running_test_pretrained_model_ddp(tmpdir):
        """Verify `test()` on pretrained model."""
        ...
        # run test set
        new_trainer = Trainer(**trainer_options)
>       new_trainer.test(pretrained_model)
tests/test_restore_models.py:60:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
pytorch_lightning/trainer/trainer.py:1189: in test
    self.run_evaluation(test_mode=True)
pytorch_lightning/trainer/evaluation_loop.py:299: in run_evaluation
    if test_mode and not self.is_overriden('test_step'):
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
self = <pytorch_lightning.trainer.trainer.Trainer object at 0x7f845ec23f90>, f_name = 'test_step', model = None
    def is_overriden(self, f_name, model=None):
        if model is None:
            model = self.get_model()
        super_object = LightningModule
        # when code pointers are different, it was overriden
>       is_overriden = getattr(model, f_name).__code__ is not getattr(super_object, f_name).__code__
E       AttributeError: 'NoneType' object has no attribute 'test_step'
pytorch_lightning/trainer/model_hooks.py:20: AttributeError
```

--- a/docs/source/weights_loading.rst
+++ b/docs/source/weights_loading.rst
@@ -5,6 +5,19 @@ Lightning can automate saving and loading checkpoints.
 
 Checkpoint saving
 -----------------
+A Lightning checkpoint has everything needed to restore a training session including:
+
+- 16-bit scaling factor (apex)
+- Current epoch
+- Global step
+- Model state_dict
+- State of all optimizers
+- State of all learningRate schedulers
+- State of all callbacks
+- The hyperparameters used for that model if passed in as hparams (Argparse.Namespace)
+
+Automatic saving
+^^^^^^^^^^^^^^^^
 
 Checkpointing is enabled by default to the current working directory.
 To change the checkpoint path pass in:
@@ -58,6 +71,15 @@ The Lightning checkpoint also saves the hparams (hyperparams) passed into the Li
 
        def __init__(self, hparams, ...):
            self.hparams = hparams
+
+Manual saving
+^^^^^^^^^^^^^
+
+To save your own checkpoint call:
+
+.. code-block:: python
+
+   model.save_checkpoint(PATH)
 
 Checkpoint Loading
 ------------------

--- a/pytorch_lightning/trainer/callback_config.py
+++ b/pytorch_lightning/trainer/callback_config.py
@@ -32,6 +32,7 @@ class TrainerCallbackConfigMixin(ABC):
         User provided weights_saved_path
         Otherwise use os.getcwd()
         """
+        ckpt_path = self.default_save_path
         if self.checkpoint_callback is True:
             # init a default one
             if self.logger is not None:
@@ -53,7 +54,8 @@ class TrainerCallbackConfigMixin(ABC):
             )
         elif self.checkpoint_callback is False:
             self.checkpoint_callback = None
-            self.ckpt_path=self.default_save_path
+
+        self.ckpt_path = ckpt_path
 
         if self.checkpoint_callback:
             # set the path for the callbacks

--- a/pytorch_lightning/trainer/callback_config.py
+++ b/pytorch_lightning/trainer/callback_config.py
@@ -12,6 +12,9 @@ class TrainerCallbackConfigMixin(ABC):
     #  the proper values/initialisation should be done in child class
     default_save_path: str
     logger: Union[LightningLoggerBase, bool]
+    weights_save_path: str
+    ckpt_path: str
+    checkpoint_callback: ModelCheckpoint
 
     @property
     @abstractmethod
@@ -44,11 +47,13 @@ class TrainerCallbackConfigMixin(ABC):
             else:
                 ckpt_path = os.path.join(self.default_save_path, "checkpoints")
 
+            self.ckpt_path = ckpt_path
             self.checkpoint_callback = ModelCheckpoint(
                 filepath=ckpt_path
             )
         elif self.checkpoint_callback is False:
             self.checkpoint_callback = None
+            self.ckpt_path=self.default_save_path
 
         if self.checkpoint_callback:
             # set the path for the callbacks

--- a/pytorch_lightning/trainer/distrib_data_parallel.py
+++ b/pytorch_lightning/trainer/distrib_data_parallel.py
@@ -145,7 +145,7 @@ class TrainerDDPMixin(ABC):
     use_amp: bool
     amp_level: str
     use_tpu: bool
-    ckpt_path: str
+    default_save_path: str
 
     @property
     @abstractmethod
@@ -350,7 +350,7 @@ class TrainerDDPMixin(ABC):
         :param model:
         :return:
         """
-        path = os.path.join(self.ckpt_path, '__temp_weight_ddp_end.ckpt')
+        path = os.path.join(self.default_save_path, '__temp_weight_ddp_end.ckpt')
         self.save_checkpoint(path)
 
     def load_spawn_weights(self, model):
@@ -360,7 +360,7 @@ class TrainerDDPMixin(ABC):
         :return:
         """
         import pdb; pdb.set_trace()
-        path = os.path.join(self.ckpt_path, '__temp_weight_ddp_end.ckpt')
+        path = os.path.join(self.default_save_path, '__temp_weight_ddp_end.ckpt')
         loaded_model = model.__class__.load_from_checkpoint(path)
         return loaded_model
 

--- a/pytorch_lightning/trainer/distrib_data_parallel.py
+++ b/pytorch_lightning/trainer/distrib_data_parallel.py
@@ -340,6 +340,11 @@ class TrainerDDPMixin(ABC):
         # continue training routine
         self.run_pretrain_routine(model)
 
+
+    def save_spawn_weights(self, model):
+        path =
+        model.save_checkpoint()
+
     def resolve_root_node_address(self, root_node):
         if '[' in root_node:
             name = root_node.split('[')[0]

--- a/pytorch_lightning/trainer/distrib_data_parallel.py
+++ b/pytorch_lightning/trainer/distrib_data_parallel.py
@@ -353,16 +353,22 @@ class TrainerDDPMixin(ABC):
         path = os.path.join(self.default_save_path, '__temp_weight_ddp_end.ckpt')
         self.save_checkpoint(path)
 
-    def load_spawn_weights(self, model):
+    def load_spawn_weights(self, original_model):
         """
         Load the temp weights saved in the process
+        To recover the trained model from the ddp process we load the saved weights
         :param model:
         :return:
         """
-        import pdb; pdb.set_trace()
+        # load weights saved in ddp
         path = os.path.join(self.default_save_path, '__temp_weight_ddp_end.ckpt')
-        loaded_model = model.__class__.load_from_checkpoint(path)
-        return loaded_model
+        loaded_model = original_model.__class__.load_from_checkpoint(path)
+
+        # copy loaded weights to old model
+        original_model.load_state_dict(loaded_model.state_dict())
+
+        # remove ddp weights
+        os.remove(path)
 
     def resolve_root_node_address(self, root_node):
         if '[' in root_node:

--- a/pytorch_lightning/trainer/distrib_data_parallel.py
+++ b/pytorch_lightning/trainer/distrib_data_parallel.py
@@ -145,6 +145,7 @@ class TrainerDDPMixin(ABC):
     use_amp: bool
     amp_level: str
     use_tpu: bool
+    ckpt_path: str
 
     @property
     @abstractmethod
@@ -340,10 +341,28 @@ class TrainerDDPMixin(ABC):
         # continue training routine
         self.run_pretrain_routine(model)
 
+        # when ddp ends, we save the model
+        self.save_spawn_weights(model)
 
     def save_spawn_weights(self, model):
-        path =
-        model.save_checkpoint()
+        """
+        Dump a temporary checkpoint after ddp ends to get weights out of the process
+        :param model:
+        :return:
+        """
+        path = os.path.join(self.ckpt_path, '__temp_weight_ddp_end.ckpt')
+        model.save_checkpoint(path)
+
+    def load_spawn_weights(self, model):
+        """
+        Load the temp weights saved in the process
+        :param model:
+        :return:
+        """
+        import pdb; pdb.set_trace()
+        path = os.path.join(self.ckpt_path, '__temp_weight_ddp_end.ckpt')
+        loaded_model = model.__class__.load_from_checkpoint(path)
+        return loaded_model
 
     def resolve_root_node_address(self, root_node):
         if '[' in root_node:

--- a/pytorch_lightning/trainer/distrib_data_parallel.py
+++ b/pytorch_lightning/trainer/distrib_data_parallel.py
@@ -351,8 +351,7 @@ class TrainerDDPMixin(ABC):
         :return:
         """
         path = os.path.join(self.ckpt_path, '__temp_weight_ddp_end.ckpt')
-        model = self.get_model()
-        model.save_checkpoint(path)
+        self.save_checkpoint(path)
 
     def load_spawn_weights(self, model):
         """

--- a/pytorch_lightning/trainer/distrib_data_parallel.py
+++ b/pytorch_lightning/trainer/distrib_data_parallel.py
@@ -351,6 +351,7 @@ class TrainerDDPMixin(ABC):
         :return:
         """
         path = os.path.join(self.ckpt_path, '__temp_weight_ddp_end.ckpt')
+        model = self.get_model()
         model.save_checkpoint(path)
 
     def load_spawn_weights(self, model):

--- a/pytorch_lightning/trainer/distrib_parts.py
+++ b/pytorch_lightning/trainer/distrib_parts.py
@@ -387,10 +387,6 @@ class TrainerDPMixin(ABC):
         """Warning: this is just empty shell for code implemented in other class."""
 
     @abstractmethod
-    def save_spawn_weights(self, *args):
-        """Warning: this is just empty shell for code implemented in other class."""
-
-    @abstractmethod
     def init_optimizers(self, *args):
         """Warning: this is just empty shell for code implemented in other class."""
 

--- a/pytorch_lightning/trainer/distrib_parts.py
+++ b/pytorch_lightning/trainer/distrib_parts.py
@@ -387,6 +387,10 @@ class TrainerDPMixin(ABC):
         """Warning: this is just empty shell for code implemented in other class."""
 
     @abstractmethod
+    def save_spawn_weights(self, *args):
+        """Warning: this is just empty shell for code implemented in other class."""
+
+    @abstractmethod
     def init_optimizers(self, *args):
         """Warning: this is just empty shell for code implemented in other class."""
 
@@ -495,6 +499,8 @@ class TrainerDPMixin(ABC):
             f'global rank: {self.tpu_global_core_rank}'
         log.info(m)
         self.run_pretrain_routine(model)
+
+        self.save_spawn_weights(model)
 
     def dp_train(self, model):
 

--- a/pytorch_lightning/trainer/evaluation_loop.py
+++ b/pytorch_lightning/trainer/evaluation_loop.py
@@ -349,7 +349,11 @@ class TrainerEvaluationLoopMixin(ABC):
 
         # log results of test
         if test_mode:
-            model.print(prog_bar_metrics)
+            if self.proc_rank == 0:
+                print('-' * 100)
+                print('TEST RESULTS')
+                print(prog_bar_metrics)
+                print('-' * 100)
 
         # log metrics
         self.log_metrics(log_metrics, {})

--- a/pytorch_lightning/trainer/trainer.py
+++ b/pytorch_lightning/trainer/trainer.py
@@ -1196,7 +1196,7 @@ class Trainer(TrainerIOMixin,
         if model is not None:
             self.model = model
             self.fit(model)
-        elif self.model is not None and self.use_ddp or self.use_tpu:
+        elif self.model is not None and (self.use_ddp or self.use_tpu):
             self.fit(self.model)
         else:
             self.run_evaluation(test_mode=True)

--- a/pytorch_lightning/trainer/trainer.py
+++ b/pytorch_lightning/trainer/trainer.py
@@ -1190,6 +1190,7 @@ class Trainer(TrainerIOMixin,
             trainer = Trainer()
             trainer.test(model)
         """
+        import pdb; pdb.set_trace()
         self.testing = True
         if model is not None:
             self.model = model

--- a/pytorch_lightning/trainer/trainer.py
+++ b/pytorch_lightning/trainer/trainer.py
@@ -745,6 +745,7 @@ class Trainer(TrainerIOMixin,
         # configure checkpoint callback
         self.checkpoint_callback = checkpoint_callback
         self.weights_save_path = weights_save_path
+        self.ckpt_path = None
 
         # accumulated grads
         self.configure_accumulated_gradients(accumulate_grad_batches)
@@ -1092,7 +1093,6 @@ class Trainer(TrainerIOMixin,
             torch_xla.core.xla_model.rendezvous("pl.Trainer.run_pretrain_routine")
 
         # set up checkpoint callback
-        self.ckpt_path = None
         self.configure_checkpoint_callback()
 
         # register auto-resubmit when on SLURM

--- a/pytorch_lightning/trainer/trainer.py
+++ b/pytorch_lightning/trainer/trainer.py
@@ -1192,11 +1192,12 @@ class Trainer(TrainerIOMixin,
             trainer = Trainer()
             trainer.test(model)
         """
-        import pdb; pdb.set_trace()
         self.testing = True
         if model is not None:
             self.model = model
             self.fit(model)
+        elif self.model is not None and self.use_ddp or self.use_tpu:
+            self.fit(self.model)
         else:
             self.run_evaluation(test_mode=True)
 

--- a/pytorch_lightning/trainer/trainer.py
+++ b/pytorch_lightning/trainer/trainer.py
@@ -745,7 +745,6 @@ class Trainer(TrainerIOMixin,
         # configure checkpoint callback
         self.checkpoint_callback = checkpoint_callback
         self.weights_save_path = weights_save_path
-        self.ckpt_path = None
 
         # accumulated grads
         self.configure_accumulated_gradients(accumulate_grad_batches)

--- a/pytorch_lightning/trainer/trainer.py
+++ b/pytorch_lightning/trainer/trainer.py
@@ -1094,7 +1094,6 @@ class Trainer(TrainerIOMixin,
         # set up checkpoint callback
         self.ckpt_path = None
         self.configure_checkpoint_callback()
-        print(self.ckpt_path)
 
         # register auto-resubmit when on SLURM
         self.register_slurm_signal_handlers()

--- a/pytorch_lightning/trainer/trainer.py
+++ b/pytorch_lightning/trainer/trainer.py
@@ -961,6 +961,7 @@ class Trainer(TrainerIOMixin,
                 self.__set_random_port()
                 mp.spawn(self.ddp_train, nprocs=self.num_gpus, args=(model,))
                 self.load_spawn_weights(model)
+                self.model = model
 
         # 1 gpu or dp option triggers training using DP module
         # easier to avoid NCCL issues
@@ -977,6 +978,7 @@ class Trainer(TrainerIOMixin,
             start_method = 'fork' if os.getenv('COLAB_GPU') else 'spawn'
             xmp.spawn(self.tpu_train, args=(model,), nprocs=self.num_tpu_cores, start_method=start_method)
             self.load_spawn_weights(model)
+            self.model = model
 
         # ON CPU
         else:

--- a/pytorch_lightning/trainer/trainer.py
+++ b/pytorch_lightning/trainer/trainer.py
@@ -1094,7 +1094,7 @@ class Trainer(TrainerIOMixin,
         # set up checkpoint callback
         self.ckpt_path = None
         self.configure_checkpoint_callback()
-        import pdb; pdb.set_trace()
+        print(self.ckpt_path)
 
         # register auto-resubmit when on SLURM
         self.register_slurm_signal_handlers()

--- a/pytorch_lightning/trainer/trainer.py
+++ b/pytorch_lightning/trainer/trainer.py
@@ -960,6 +960,7 @@ class Trainer(TrainerIOMixin,
             else:
                 self.__set_random_port()
                 mp.spawn(self.ddp_train, nprocs=self.num_gpus, args=(model,))
+                self.load_spawn_weights(model)
 
         # 1 gpu or dp option triggers training using DP module
         # easier to avoid NCCL issues
@@ -975,6 +976,7 @@ class Trainer(TrainerIOMixin,
             #  COLAB_GPU is an env var available by default in Colab environments.
             start_method = 'fork' if os.getenv('COLAB_GPU') else 'spawn'
             xmp.spawn(self.tpu_train, args=(model,), nprocs=self.num_tpu_cores, start_method=start_method)
+            self.load_spawn_weights(model)
 
         # ON CPU
         else:
@@ -1090,6 +1092,7 @@ class Trainer(TrainerIOMixin,
             torch_xla.core.xla_model.rendezvous("pl.Trainer.run_pretrain_routine")
 
         # set up checkpoint callback
+        self.ckpt_path = None
         self.configure_checkpoint_callback()
 
         # register auto-resubmit when on SLURM

--- a/pytorch_lightning/trainer/trainer.py
+++ b/pytorch_lightning/trainer/trainer.py
@@ -1094,6 +1094,7 @@ class Trainer(TrainerIOMixin,
         # set up checkpoint callback
         self.ckpt_path = None
         self.configure_checkpoint_callback()
+        import pdb; pdb.set_trace()
 
         # register auto-resubmit when on SLURM
         self.register_slurm_signal_handlers()

